src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java

```diff
@@ -0,0 +1,446 @@
+import net.minecraftforge.common.capabilities.Capability;
+import net.minecraftforge.common.capabilities.ForgeCapabilities;
+import net.minecraftforge.common.capabilities.ICapabilityProvider;
+import net.minecraftforge.common.util.LazyOptional;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+import net.minecraftforge.items.IItemHandler;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+public class InterfaceCapabilityProviderMapper implements CapabilityProviderMapper {
+    @Override
+    public Optional<ICapabilityProvider> getProviderFor(LevelAccessor level, BlockPos pos) {
+        var be = level.getBlockEntity(pos);
+        if (!(be instanceof InterfaceBlockEntity in)) {
+            return Optional.empty();
+        }
+
+        if (!in.getConfig().isEmpty() || in.getMainNode() == null || in.getGridNode() == null || !in.getGridNode().isActive()) {
+            return Optional.empty();
+        }
+
+        var cap = be.getCapability(Capabilities.STORAGE_MONITORABLE_ACCESSOR);
+        if (!cap.isPresent()) {
+            return Optional.empty();
+        }
+
+        return Optional.of(new InterfaceCapabilityProvider(level, pos));
+    }
+
+    private record InterfaceCapabilityProvider(LevelAccessor level, BlockPos pos) implements ICapabilityProvider {
+        @Override
+        public @NotNull <T> LazyOptional<T> getCapability(@NotNull Capability<T> cap, @Nullable Direction side) {
+            if (cap == ForgeCapabilities.ITEM_HANDLER || cap == ForgeCapabilities.FLUID_HANDLER) {
+                return LazyOptional.of(() -> new InterfaceHandler(level, pos)).cast();
+            }
+
+            var in = interfaceAt(level, pos);
+            if (in != null) {
+                return in.getCapability(cap, side);
+            }
+
+            return LazyOptional.empty();
+        }
+    }
+
+    @Nullable
+    private static InterfaceBlockEntity interfaceAt(LevelAccessor level, BlockPos pos) {
+        var be = level.getBlockEntity(pos);
+        if (be instanceof InterfaceBlockEntity in) {
+            return in;
+        }
+        return null;
+    }
+
+    record InterfaceHandler(LevelAccessor level, BlockPos pos) implements IItemHandler, IFluidHandler {
+        @Nullable
+        InterfaceBlockEntity getInterface() {
+            return interfaceAt(this.level, this.pos);
+        }
+
+        @Nullable
+        IEnergyService getEnergy() {
+            var in = this.getInterface();
+            if (in == null) {
+                return null;
+            }
+
+            var grid = in.getMainNode().getGrid();
+            if (grid == null) {
+                return null;
+            }
+
+            return grid.getEnergyService();
+        }
+
+        @NotNull
+        LazyOptional<IStorageMonitorableAccessor> getCapability() {
```

Comment:
```suggestion
        LazyOptional<IStorageMonitorableAccessor> getCapability() {
```
the package file should be adding notnullbydefault to everything